proxy/api/src/http/notification.rs

```diff
@@ -0,0 +1,86 @@
+//! Unidirectional stream of events happening in the proxy. This enables exposing tailing logs to
+//! users, or widgets which show topology information like how many and what peers are connected.
+
+use warp::filters::BoxedFilter;
+use warp::{Filter, Reply};
+
+use crate::notification::Subscriptions;
+
+/// SSE based notifications endpoint.
+pub fn filters(subscriptions: Subscriptions) -> BoxedFilter<(impl Reply,)> {
+    stream_filter(subscriptions)
+}
+
+/// `GET /`
+pub fn stream_filter(subscriptions: Subscriptions) -> BoxedFilter<(impl Reply,)> {
+    warp::get()
+        .and(warp::any().map(move || subscriptions.clone()))
+        .and_then(handler::stream)
+        .boxed()
+}
+
+/// Notification handlers to serve event streams.
+mod handler {
+    use std::convert::Infallible;
+
+    use futures::{Stream, StreamExt as _};
+    use tokio::sync::mpsc;
+    use warp::{sse, Rejection, Reply};
+
+    use crate::notification::{Notification, Subscriptions};
+
+    /// Provides an SSE events endpoint providing a stream of
+    /// [`crate::notification::Notification`]s.
+    pub async fn stream(subscriptions: Subscriptions) -> Result<impl Reply, Rejection> {
+        let subscriber = subscriptions.subscribe().await;
+
+        Ok(sse::reply(sse::keep_alive().stream(map(subscriber))))
+    }
+
+    /// Maps the [`crate::notification::Notification`] from a subscription to an
+    /// [`sse::ServerSideEvent`].
+    fn map(
+        subscriber: mpsc::UnboundedReceiver<Notification>,
+    ) -> impl Stream<Item = Result<impl sse::ServerSentEvent, Infallible>> {
+        subscriber.map(|notification| match notification {
+            Notification::LocalPeerListening(addr) => {
+                Ok((sse::event("LOCAL_PEER_LISTENING"), sse::json(addr)))
```

Comment:
To keep things in one place, could we have a method on `Notification` that either returns an `sse::event` or a `&'static str`?

Reply:
Was about to do that, but there are a couple of oddities. Notifications shouldn't know about its usage as sse event. Was thinking about impling `Display`, but it would be a bit whack if it just renders the enum name. What do you suggest?

Reply:
Coud you implement an `Into<sse::Event> for Notification` inside the `http/notifications` module?

Reply:
Wouldn't "keep things in one place" but at least there would be only one way of constructing an event from it.

Reply:
It would keep "things in the right place" as it would keep the concerns of how to construct the sse events in the context where they are used. Dunno if the `Into` is possible as `sse::SeverSentEvent` (what `sse::event` returns) is a trait.